server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -371,7 +371,7 @@ public final Collection<IgnoredSourceFieldMapper.NameValue> getIgnoredFieldsMiss
      * Applies to synthetic source only.
      */
     public final DocumentParserContext maybeCloneForArray(Mapper mapper) throws IOException {
-        if (canAddIgnoredField() && mapper instanceof ObjectMapper) {
+        if (canAddIgnoredField() && mapper instanceof ObjectMapper && indexSettings().isSyntheticSourceSecondDocParsingPassEnabled()) {
```

Comment:
Can we avoid calling this (reading a volatile field) per `array` field and do it once per document instead? I'm benchmarking the impact of https://github.com/elastic/elasticsearch/blob/363143fa39822300534a8d4f80aa267c6b2ede59/server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java#L444.

Reply:
Interesting, I didn't realize that volatiles are so expensive.. Done.